sonic-net/sonic-utilities
# CLI support for SmartSwitch PMON

#### What I did
Enhanced the following CLIs to support SmartSwitch PMON as described in the PMON HLD documentation "https://github.com/sonic-net/SONiC/blob/d19d8933a43d0a31a4f3b2310f4336f289bca340/doc/smart-switch/pmon/smartswitch-pmon.md"

CLIs:
Added new module "DPUX" support for 1 and 2 below
    1. "config chassis module startup DPUX" , where X could be 0, to the maximum number of DPUs-1 in the SmartSwitch chassis 
    2. "config chassis module shutdown DPUX"

Extended the following CLIs to support the new module "DPUX" and also provided the "all" option to display the reboot-cause of the "NPU" and all "DPU" modules.
    1. "show reboot-cause" will remain the same and added "show reboot-cause all"
    2.  "show reboot-cause history" will remain the same and added "show reboot-cause history \<module-name\>", where module name could be DPUX, and all.

Added a new sub command "show system-health dpu \<module-name\>", where module name could be DPUX, and all.  This new subcommand will provide additional DPU state details as mentioned in the HLD

#### How I did it
1. Kept the original CLI output unaltered
2. Added sub command to support SmartSwitch "DPUs"
3. Added additional code in chassisd, and in platform modules.py, chassis.py to support it
4. Updated the DB tables as mentioned in the PMON HLD

#### Fixes: https://github.com/sonic-net/sonic-buildimage/issues/21372

#### How to verify it
1. Build an image with the required files (refer to the other upstream PRs and the platform PRs)
    Require files:
        - This PR including reboot_cause.py, chassis_modules.py, system_health.py)
        - The other PR including module_base.py, chassis_base.py, docker-pmon.supervisord.conf.j2, chassisd, mock_module_base.py, and the appropriate database_config.json
        - Platform "platform-cisco-8000"  supporting PMON (module.py, chassis.py, inventory.py, pmon_daemon_control.json, and the required DB changes)
2. Run the CLIs and see the new output

#### CLI output
root@sonic:~# show reboot-cause
Unknown

root@MtFuji:/home/cisco# show reboot-cause all
Device    Name                 Cause       Time                             User
--------  -------------------  ----------  -------------------------------  ------
NPU       2024_12_11_01_54_02  reboot      Wed Dec 11 01:48:07 AM UTC 2024  cisco
DPU0      2024_12_11_01_56_45  reboot     Wed Dec 11 01:56:45 AM UTC 2024


cisco@MtFuji-dut:~$ show reboot-cause history
Name                 Cause    Time                             User    Comment
-------------------  -------  -------------------------------  ------  ---------
2024_12_19_14_00_33  reboot   Thu Dec 19 01:55:41 PM UTC 2024  cisco   N/A
2024_12_15_08_21_38  reboot   Sun Dec 15 08:16:36 AM UTC 2024  cisco   N/A

cisco@MtFuji-dut:~$ show reboot-cause history all
Device    Name                 Cause         Time                             User    Comment
--------  -------------------  ------------  -------------------------------  ------  -------------------
NPU       2024_12_19_14_00_33  reboot        Thu Dec 19 01:55:41 PM UTC 2024  cisco   N/A
NPU       2024_12_15_08_21_38  reboot        Sun Dec 15 08:16:36 AM UTC 2024  cisco   N/A
DPU1      2024_12_19_14_36_47  Non-Hardware  Thu Dec 19 02:36:47 PM UTC 2024          Switch rebooted DPU
DPU0      2024_12_19_14_03_24  reboot       Thu Dec 19 02:03:24 PM UTC 2024          N/A
DPU0      2024_12_17_15_57_29   reboot       Tue Dec 17 03:57:29 PM UTC 2024          N/A

root@sonic:~# show reboot-cause history DPU0
Device    Name                 Cause         Time                             User    Comment
--------  -------------------  ------------  -------------------------------  ------  -------------------
DPU0      2024_12_19_14_03_24  reboot       Thu Dec 19 02:03:24 PM UTC 2024          N/A
DPU0      2024_12_17_15_57_29   reboot       Tue Dec 17 03:57:29 PM UTC 2024          N/A
